components/brave_wallet/common/zcash_utils.cc

```diff
@@ -5,18 +5,93 @@
 
 #include "brave/components/brave_wallet/common/zcash_utils.h"
 
-#include <optional>
+#include <algorithm>
+#include <string>
+#include <utility>
+#include <vector>
 
+#include "base/big_endian.h"
+#include "base/types/expected.h"
 #include "brave/components/brave_wallet/common/btc_like_serializer_stream.h"
 #include "brave/components/brave_wallet/common/encoding_utils.h"
+#include "brave/components/brave_wallet/common/f4_jumble.h"
 #include "brave/components/brave_wallet/common/hash_utils.h"
 #include "brave/third_party/bitcoin-core/src/src/base58.h"
+#include "brave/third_party/bitcoin-core/src/src/bech32.h"
+#include "brave/third_party/bitcoin-core/src/src/util/strencodings.h"
+#include "components/base32/base32.h"
 
 namespace brave_wallet {
 
 namespace {
+constexpr size_t kPaddedHrpSize = 16;
 constexpr size_t kPubKeyHashSize = 20;
 constexpr size_t kPrefixSize = 2;
+
+// https://zips.z.cash/zip-0316#encoding-of-unified-addresses
+enum AddrType {
+  P2PKH = 0x00,
+  P2PSH = 0x01,
+  Sapling = 0x02,
+  Orchard = 0x03,
+  Last = 0x04
+};
+
+using ParsedAddress = std::pair<AddrType, std::vector<uint8_t>>;
+
+std::optional<uint64_t> ReadCompactSize(base::span<const uint8_t>& data) {
+  uint64_t value;
+  if (data.size() == 0) {
+    return std::nullopt;
+  }
+  uint8_t type = data[0];
+  if (data.size() > 0 && data[0] < 253) {
```

Comment:
this could use some comments, is this format documented somewhere? Link?

Reply:
https://btcinformation.org/en/developer-reference#compactsize-unsigned-integers

Reply:
Done

Reply:
Like Brian mentioned I feel we should have better comments on why we have all these `if`/`else` blocks, rather than just a link.

Reply:
Done